[RW-1948][risk=low] Implement initial BigQuery OMOP -> Elastic indexer

calbach opened the pull request:
Ticket: https://precisionmedicineinitiative.atlassian.net/browse/RW-1948
Design: https://docs.google.com/document/d/1N_TDTOi-moTH6wrXn1Ix4dwUlw4j8GT9OsL9yXYXYmY/edit

# Overview
Command-line tool for indexing an OMOP BigQuery dataset into Elasticsearch. Executes the following pipeline:

- Create (and optionally replace) an existing Elasticsearch index for this CDR.
- Query BigQuery to produce JSON documents: the output of the query is the exact format we define for Elasticsearch (using a repeated STRUCT for repeated nested documents).
- [Optional] Export these results to a temporary table, and export that to GCS. Slower, but necessary for larger datasets (e.g. the entire Synthetic CDR).
- Bulk ingest the JSON documents into Elasticsearch (either inserts or updates).

The indexer can optionally be run on a subsample of participants in the CDR for faster test iterations and local development. The full index will be 10-100GB, so it is not advisable to run this locally against the entire synthetic CDR.

# Limitations
- Currently only indexes the person and condition_occurrence table.
- `./project.rb create-local-es-index` interface is limited and hardcoded to only work with the synthetic CDR -> local

<!--
Replace this this template with your PR description.
Please remember to keep in mind the security levels outlined in [CONTRIBUTING.md](CONTRIBUTING.md) and to 
include a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title

* **no**: None 
* **low**: Low chance of potential impact to, or exposure of patient data
* **moderate**: Moderate chance of potential impact to, or exposure of patient data
* **severe**: Severe chance of potential impact to, or exposure of patient data

Please also:

* Get thumbs from reviewer(s)
* Verify all tests go green, including CI tests
-->

gjuggler commented on api/libproject/devstart.rb:
```diff
@@ -1136,6 +1136,46 @@ def list_clusters(cmd_name, *args)
 
+def create_local_es_index(cmd_name, *args)
+  op = WbOptionsParser.new(cmd_name, args)
+  op.opts.inverse_prob = 10000
```
nitpick: actually default to 1 before this goes in.

gjuggler replied:
Makes sense – in that case, update the option description to match.

calbach replied:
Oops, no wonder you asked for the change. Done.

gjuggler commented on api/tools/build.gradle:
```diff
@@ -52,6 +52,15 @@ task manageClusters(type: JavaExec) {
 
+task elasticSearchIndexer(type: JavaExec) {
+  classpath sourceSets.main.runtimeClasspath
+  main = "org.pmiops.workbench.tools.elastic.BigQueryOmopIndexer"
```
nitpick: would it be possible to aim for consistent naming of the task / command-line / Java class? The other tools were mostly good about this.

gjuggler commented on api/tools/src/main/java/org/pmiops/workbench/tools/elastic/BigQueryOmopIndexer.java:
```diff
@@ -0,0 +1,325 @@
+ */
+public final class BigQueryOmopIndexer {
+
+  private static Option queryProjectIdOpt = Option.builder()
```
Would it be helpful / possible to add types to these options? It seems like the option API has a setType, but I have no idea if it works well and/or parses things as you'd expect. Booleans seem the most relevant.

calbach replied:
Tried this out. It doesn't seem to do anything unless you use the `getParsedOptionValue()` method: https://commons.apache.org/proper/commons-cli/javadocs/api-1.3.1/org/apache/commons/cli/CommandLine.html

This API is kind of lame as it throws a parsing Exception and only supports returning an Object which I have to cast anyways, so I think it's probably clearer if I just parse it myself. Overall not really impressed with this package so far..

Also booleans are straightforward, they have no argument, so we just check for presence (there is no negation flag built in)

gjuggler commented on api/tools/src/main/java/org/pmiops/workbench/tools/elastic/ElasticDocument.java:
```diff
@@ -0,0 +1,175 @@
+   */
+  public static ElasticDocument fromBigQueryJson(String line) throws IOException {
+    JsonObject doc = new JsonParser().parse(line).getAsJsonObject();
+    String id = doc.remove("_id").getAsString();
```
This code doesn't have tests, so liberally commenting important-but-obscure lines is more important. You'd commented on this issue (removing _id_ from the doc) in the other file, but it's warranted here too.

... but then again, it may also be worth adding basic unit tests on this class at some point. Seems reasonably straightforward to do.

calbach replied:
Yes, I'm thinking about this. Unfortunately to write a meaningful test here, I'll need BigQuery and/or Elastic. It seems there is still no suitable fake for BigQuery locally aside from mocking out its exact behavior (a lot of work, and critical behavior we'd want to test). I'm thinking I might compromise by stubbing the GCS piece with golden data (saved BigQuery output), then running this tool via our "api-local-test" (runs on each master commit / PR), which starts up the API server, then tests. Another alternative would be to write an integration test, but I'd like some more immediate coverage here than "nightly".

Any thoughts?

freemabd1 replied:
This could be helpful: https://github.com/dadoonet/testcontainers-java-module-elasticsearch

freemabd1 commented on api/tools/src/main/java/org/pmiops/workbench/tools/elastic/BigQueryOmopIndexer.java:
```diff
@@ -0,0 +1,325 @@
+        .getService();
+    String cdrDataset = opts.getOptionValue(cdrBigQueryDatasetOpt.getLongOpt());
+    int totalSampleSize = getPersonCount(bq, cdrDataset, inverseProb);
+    String personSQL = getPersonBigQuerySQL(cdrDataset, inverseProb);
```
can the count not be gotten from the result of personSql?

calbach replied:
I'd assumed I couldn't get the total row count when I did an extraction to a table; I was wrong. Done - thanks.

freemabd1 commented on api/tools/src/main/java/org/pmiops/workbench/tools/elastic/BigQueryOmopIndexer.java:
```diff
@@ -0,0 +1,325 @@
+        new HttpHost(url.getHost(), url.getPort(), url.getProtocol())));
+  }
+
+  @Bean
```
Do we need to expose this as a Spring Bean?

calbach replied:
Tested - yes, this is what triggers the SpringApplicationBuilder to actually execute something.

freemabd1 commented on api/tools/src/main/resources/bigquery/es_person.sql:
```diff
@@ -0,0 +1,56 @@
+SELECT
```
Is this sql going to be a template for all domains? If so should you parameterized anything that references a domain? or this can be done later as this code matures.

calbach replied:
I'm planning to throw the whole mess into here in a subsequent PR as one giant query. I think each domain will have slightly different values being mapped into these structs.

calbach commented on api/tools/src/main/java/org/pmiops/workbench/tools/elastic/BigQueryOmopIndexer.java:
```diff
@@ -0,0 +1,325 @@
+public final class BigQueryOmopIndexer {
+
+  private static Option queryProjectIdOpt = Option.builder()
+      .longOpt("query-project-id")
```
I'm effectively using this Option as the constant elsewhere in the file. I think this is probably overall lower boilerplate.

calbach commented on api/tools/src/main/java/org/pmiops/workbench/tools/elastic/BigQueryOmopIndexer.java:
```diff
@@ -0,0 +1,325 @@
+      .required()
+      .hasArg()
+      .build();
+  private static Option esBaseUrlOpt = Option.builder()
```
I think I prefer the bulider form since some of these need slightly different modifiers. I did a few iterations of this and I landed on this as an "OK" solution here. The thing I'd lose if I used that approach (if I'm understanding correctly), is the ability to reference the options throughout the code. I initially was retyping the flag strings, which was error-prone. Moving those strings to constants feels like higher boilerplate than this approach to me, though that might be wrong.

calbach commented:
I started looking at adding tests here. I think I'll wait for @gjuggler 's CLI PR to go in and switch to his approach so I can avoid duplicating effort in setting up the test gradle deps here.

calbach commented:
Updated to use @gjuggler 's new gradle setup. I also refactored a few things to make it a bit more conducive to future test writing.

After mulling on this a bit, I think this tool would be best covered by an "integration" test - specifically a test harness which has an elastic docker image wired in via our docker-compose setup. I tried out the testcontainers stuff @freemabd1 linked - but it doesn't seem to be easily compatible with our current docker setup, e.g. for CircleCI they recommend doing something pretty custom: https://www.testcontainers.org/supported_docker_environment/continuous_integration/circle_ci/. I'm not yet sure whether that integration test would hit live BigQuery - but it might. As this will be a larger change, my preference would be to handle this as part of this ticket: https://precisionmedicineinitiative.atlassian.net/browse/RW-1957

Does this seem reasonable?

There's also a piece of work involved here which is figuring out how to incorporate tools tests into the gradle setup. I tried a couple of approaches without much immediate success.
